codice/ddf
# DDF-4642 Update FeatureMetacardType to handle simpleTypes and simpleContent to arbitrary depths

#### What does this PR do?

Previously, when evaluating a feature property whose type is a simpleType, FeatureMetacardType would only go one level deep to determine its ‘base’ type. The ‘base’ type is used to determine what
AttributeType to use to map to the feature property, e.g. string, date, or integer. If the ‘base’ type is a type we can map to an AttributeType then FeatureMetacardType adds the feature property to itself as an attribute descriptor. Only going one level deep meant that feature properties whose type was a ‘base’ schema type such as xsd:string or xsd:int would be mapped and added as attribute descriptor, but feature properties whose type was a custom-defined simpleType extending a ‘base’ type would not be because we’d look at its QName, see it’s not a ‘base’ type, and pass over it. As a result, feature properties whose type was a simpleType were not queryable at all.

This commit changes how FeatureMetacardType handles simpleTypes. Now it will attempt to find a simpleType’s root type by following <list> and <restriction> definitions until a ‘base’ schema type is reached.

This commit also updates FeatureMetacardType to handle simpleContent inside complexTypes. Now simpleContent `<restriction>`s and `<extension>`s are handled in virtually the same way simpleTypes are handled as described above.

The upshot is that WFS sources can now query feature properties whose type is a simpleType or a complexType with simpleContent and that type’s root type is a ‘base’ schema type that DDF can understand.

The three FeatureMetacardTypeTest classes in their respective WFS 1.0.0, 1.1.0, and 2.0.0 modules have been consolidated into one FeatureMetacardTypeTest that lives in the spatial-wfs-common module. FeatureMetacardType lives in the spatial-wfs-common module so the unit tests for it in the other three modules did not count for unit test coverage against FeatureMetacardType. Moreover, the tests in the three classes were virtually identical so consolidating them into one made sense.

Finally, the XmlSchema object used in the FeatureMetacardType unit tests is now obtained by reading a real schema file instead of being constructed manually. The schema object trees constructed in the tests did not accurately model the schema object trees seen at runtime when a real schema file is parsed. Now that the unit tests obtain the XmlSchema object in the same way the runtime code does, they will be identical.

#### Who is reviewing it? 
@glenhein 
@leo-sakh 
@Corey-Collins 

#### Select relevant component teams: 
@codice/io 
@codice/ogc 

#### Ask 2 committers to review/merge the PR and tag them here.
@bdeining
@troymohl

#### How should this be tested?
1. Create a WFS 1.1.0 source with the URL http://services.azgs.az.gov/ArcGIS/services/OneGeology/AZGS_HIGS_Geology_WFS/MapServer/WFSServer
a. Disable the CN check and allow redirects. Leave everything else as-is.
2. In the Karaf console, run `log:set info org.apache.cxf` and `log:set warn org.apache.camel`.
3. Open Intrigue.
4. Run an anyText * query on the source. It may take a while to complete.
5. Check DDF's logs for an `Outbound Message` to http://services.azgs.az.gov/ArcGIS/services/OneGeology/AZGS_HIGS_Geology_WFS/MapServer/WFSServer. The request body should be a `GetFeature` XML request.
6. Open http://services.azgs.az.gov/ArcGIS/services/OneGeology/AZGS_HIGS_Geology_WFS/MapServer/WFSServer?SERVICE=WFS&REQUEST=DescribeFeatureType.
7. Check the GetFeature XML in the console and verify that all of the `element`s inside `GeologicUnitViewType` from step 6 _except_ `shape`, `OBJECTID`, `shape_Length`, and `shape_Area` are present in the query.
a. These elements are all custom-defined `simpleType`s that would not have been included in the query without this change.

#### Any background context you want to provide?
#### What are the relevant tickets?
Fixes: #4642 

#### Screenshots

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
